### PR TITLE
Fix state-caching 

### PR DESCRIPTION
**Pull request by richardpringle**
State-caching appears to have always been a little broken. While fixing this, I also realized that the `state_key` macro was doing a bunch of things that didn't make any sense, so I'm also fixing that since I have to change it anyway.

### CODE REVIEW

File: x/programs/rust/sdk-macros/src/lib.rs [hunk 1]
```diff
@@ -40,6 +40,22 @@ pub fn public(_: TokenStream, item: TokenStream) -> TokenStream {
...
+                            .last_mut()
+                            .map(|segment| &mut segment.arguments)
+                    });
+
+                    if let [Some(context_type_args), Some(ty_args)] = args {
+                        *context_type_args = ty_args.clone();
+                    }
+                }
+
                 std::mem::swap(&mut context_type, ty);
```
Comment (iFrostizz) [inline x/programs/rust/sdk-macros/src/lib.rs:59]: I think that you can replace all of this by

```suggestion
                context_type = ty.clone();
```

That is going to preserve the explicit generic passed to the `Context` struct if it is needed to not be `()`

**Review by iFrostizz — APPROVED** [review]

File: x/programs/rust/sdk-macros/src/lib.rs [hunk 2]
```diff
@@ -179,74 +195,75 @@ pub fn public(_: TokenStream, item: TokenStream) -> TokenStream {
...
+            Fields::Unnamed(_) => quote! {
+                Self::#variant_ident(a) => wasmlanche_sdk::state::PrefixedBytes::new(#idx, a.as_ref())
+            },
+            // ex: Point
+            Fields::Unit => quote! {
+                Self::#variant_ident => wasmlanche_sdk::state::PrefixedBytes::new(#idx, &[])
+            },
+            // ex: Point { x: f64, y: f64 }
+            Fields::Named(_) => quote! {
+                Self::#variant_ident { .. } => panic!("named enum fields are not supported")
```
[the remark this answers is not in the document]
Comment (richardpringle) [reply x/programs/rust/sdk-macros/src/lib.rs:233]: Woooooow, I was just moving this, I didn't change it. You're absolutely correct, good catch!

File: x/programs/rust/sdk-macros/src/lib.rs [hunk 3]
```diff
@@ -179,74 +195,75 @@ pub fn public(_: TokenStream, item: TokenStream) -> TokenStream {
...
+
+        if let Some(segment) = type_path.path.segments.last() {
+            let context_type = context_type.segments.last_mut();
+            context_type.unwrap().arguments = segment.arguments.clone();
+        }
+
+        type_path.path.segments.last() == context_type.segments.last()
     } else {
         false
     }
```
[the remark this answers is not in the document]
Comment (richardpringle) [reply x/programs/rust/sdk-macros/src/lib.rs:269]: I would approve a PR with this haha. I don't quite think it's necessary right now because the `ui` tests cover the cases pretty well, but more unit tests usually make things easier to expand and extend.

File: x/programs/rust/wasmlanche-sdk/src/state.rs [hunk 4]
```diff
@@ -38,56 +38,40 @@ pub enum Error {
     Delete,
 }
 
-pub struct State<K>
-where
-    K: Into<Key> + Hash + PartialEq + Eq + Clone,
-{
-    cache: HashMap<K, Vec<u8>>,
+pub struct State<'a, K: Key> {
+    cache: &'a RefCell<HashMap<K, Vec<u8>>>,
```
[the remark this answers is not in the document]
Comment (richardpringle) [reply x/programs/rust/wasmlanche-sdk/src/state.rs:42]: If programs were multi-threaded, I think it would be a lot harder to guarantee determinism. I don't think it's something we're going to have to worry about in the near future. 

Also, `RefCell` isn't `Sync`, you can still send it across threads
https://doc.rust-lang.org/stable/std/cell/struct.RefCell.html#impl-Send-for-RefCell%3CT%3E

File: x/programs/rust/sdk-macros/src/lib.rs [hunk 5]
```diff
@@ -179,74 +195,75 @@ pub fn public(_: TokenStream, item: TokenStream) -> TokenStream {
...
 
 /// Returns whether the type_path represents a Program type.
 fn is_context(type_path: &std::boxed::Box<Type>) -> bool {
     if let Type::Path(type_path) = type_path.as_ref() {
-        type_path.path.segments.last() == parse_str::<Path>(CONTEXT_TYPE).unwrap().segments.last()
+        let mut context_type = parse_str::<Path>(CONTEXT_TYPE).unwrap();
+
+        if let Some(segment) = type_path.path.segments.last() {
+            let context_type = context_type.segments.last_mut();
+            context_type.unwrap().arguments = segment.arguments.clone();
```
[the remark this answers is not in the document]
Comment (richardpringle) [reply x/programs/rust/sdk-macros/src/lib.rs]: That unwrap was a little confusing so I moved it up a line. This isn't a `context_generic` though. So I don't know what the generic is since it's input by the user, so I obviously can't use it in the `is_context` function. 

However, I'm not just realizing that I can just compare the `ident`. Will change...

**Comment by dboehm-avalabs** [comment]
Is splitting the functions into "func" and "func_internal" needed?  It seems a little messy, so it would be nice if that weren't necessary.

**Review by dboehm-avalabs — APPROVED** [review]

**Comment by richardpringle** [comment]
> Is splitting the functions into "func" and "func_internal" needed? It seems a little messy, so it would be nice if that weren't necessary.

TL;DR Response: 
I agree, but there's some functionality I want to get in before dealing with the cosmetics.

I agree it's a little messy for now. I want to go in and generate the code for program-to-program calls (#723) then and maybe get #595 in as well, then take a look at the API and calling conventions and see what needs cleaning up. 

I have a couple of different ideas on how to deal with that `func_internal`, but I don't think they are worth changing now in case other API changes are necessary for functionality.

File: x/programs/rust/examples/token/src/lib.rs [hunk 6]
```diff
@@ -19,94 +19,109 @@ enum StateKey {
...
 #[public]
-pub fn mint_to(context: Context, recipient: Address, amount: i64) -> bool {
-    let Context { program, .. } = context;
+pub fn mint_to(context: Context<StateKeys>, recipient: Address, amount: i64) -> bool {
+    mint_to_internal(context, recipient, amount);
+    true
+}
+
+fn mint_to_internal(
+    context: Context<StateKeys>,
```
[the remark this answers is not in the document]
Comment (richardpringle) [reply x/programs/rust/examples/token/src/lib.rs:62]: Yeah, I agree. This isn't going to stay like this though. 
See the comment [above](https://github.com/ava-labs/hypersdk/pull/916#issuecomment-2124882865)
